refactor: developer tools.

NoSharp opened the pull request:
This is a near complete refactor of how the previous iteration of this developer tools utility that was made.

**Please test this properly before deploying to production. I cannot test this on my single player game.**

NoSharp committed:
refactor: optimise and improve redability on "recureRefresh"

NoSharp committed:
refactor: clean up devtools a little bit more.

Use CompileFile over RunString so we can
allow for some room to sandbox it more later on.

There could be some more function localisation done.
However there will be a point where it becomes unreadable.

NoSharp committed:
refactor: massively overhaul the layout of the code.

TomDotBat commented on lua/autorun/client/cl_devtools.lua:
```diff
@@ -1,144 +1,165 @@
 SWAMP_DEV = SWAMP_DEV or {}
-SWAMP_DEV.isDev = SWAMP_DEV.isDev
+-- Are we in development mode?
+SWAMP_DEV.isDev = false
```
Could just do a nil check, no real need for the assignment to false here.

NoSharp replied:
Addressed in a more recent comment.

TomDotBat commented on lua/autorun/client/cl_devtools.lua:
```diff
@@ -1,144 +1,165 @@
+        local code = file.Read(filePath, "MOD")
+        if not code then error("No file exists with the file path: " .. filePath) end
+        environmentVars = environmentVars or {}
+        loadFunc = CompileString(code, "SwapDevTools/" .. filePath)
```
Swamp*

TomDotBat commented on lua/autorun/client/cl_devtools.lua:
```diff
@@ -1,144 +1,165 @@
+        loadFunc = CompileString(code, "SwapDevTools/" .. filePath)
+        -- setup our environment
+        local environment = _G
+        for key,var in pairs(environmentVars) do
```
Would be nice to keep spacing consistency between args/for loop vars.

swampservers replied:
We automatically glualint all the files, so it's not really necessary to enforce perfect formatting

TomDotBat commented on lua/autorun/client/cl_devtools.lua:
```diff
@@ -1,144 +1,165 @@
+        debug.setfenv(loadFunc,environment)
+
+        xpcall(loadFunc, function(err)
+            print("[contrib] Error during loading file with devtools:", err)
```
Potentially not wrong - but check if this follows the common logging format in this project.

swampservers replied:
There's not really a standard

TomDotBat replied:
> There's not really a standard

Would you say this is fine as is then?

TomDotBat commented on lua/autorun/client/cl_devtools.lua:
```diff
@@ -1,144 +1,165 @@
 
 -- toggle dev editing. Root is addons/contrib
 concommand.Add("dev", function()
     if (LocalPlayer():GetRank() <= 0) then return end
```
Confirm that this is actually not going to open up dev tools to donators or something.

swampservers replied:
Rank is only for staff. However I was actually thinking we should have a hardcoded steamID list in here so devs don't have to be staff to use these tools. Then i'd update the readme to say "Make a PR to add your name to the developer list" which is cool because if someone is incapable of that they are obviously not worth spending any time helping.

TomDotBat replied:
Maybe it'd be better to put this table in a different location and made global so we can reference it in other systems. Perhaps if you wanted a contributor tag on player overheads or something, etc.

TomDotBat commented on lua/autorun/client/cl_devtools.lua:
```diff
@@ -1,144 +1,165 @@
-            SWAMP_DEV.structTypes[SWAMP_DEV.itemStruct](parentFolder)
-        else
-            SWAMP_DEV.structTypes[SWAMP_DEV.itemStruct](string.TrimRight(string.gsub(path, "^.*/", ""), ".lua"))
+local rootDirectory = "addons/contrib/"
```
As this is constant consider using UPPER_SNAKE_CASE.

TomDotBat reviewed (changes requested):
Looks alright to me - just a few small changes and verification needed from @swampservers.

swampservers commented on lua/autorun/client/cl_devtools.lua:
```diff
@@ -1,144 +1,165 @@
     ["weapons"] = function(curClass)
-        SWEP = weapons.GetStored(curClass)
+        local found = weapons.GetStored(curClass)
+        if not found then return end
```
I'd prefer like `found and {SWEP=found} or nil`

NoSharp committed:
refactor: DevTool utility

rootDirectory -> ROOT_DIRECTORY
remove assingment of SWAMP_DEV.isDev (as we null check instead)

swampservers commented on lua/autorun/client/cl_devtools.lua:
```diff
@@ -1,144 +1,165 @@
-
-local function recurseRefresh(root, times, curPath)
-    local files, folders = file.Find(root .. curPath .. "*", "MOD")
+    local fileTime = file.Time
```
What's the point of this?

NoSharp replied:
Localization reduces indexing times, as this is called frequently this will be slightly more efficient.

swampservers replied:
I really don't think the performance impact will matter. Code is cleaner without it.

swampservers replied:
You can leave it in if you want, but I think the overwhelming majority of the lag from this function will be due to the file.Time call looking at the disk synchronously.

If you're really worried about performance, what you could do is keep track of the 5 most recently edited files and update them 3x per second, and then update other files every 5 seconds, and additionally, don't do them all at once, spread them out over the 5 seconds and do a few every frame.

swampservers commented on lua/autorun/client/cl_devtools.lua:
```diff
@@ -1,144 +1,165 @@
 SWAMP_DEV = SWAMP_DEV or {}
-SWAMP_DEV.isDev = SWAMP_DEV.isDev
+-- Are we in development mode?
+SWAMP_DEV.isDev = false
```
"enabled" would make more sense to name this

NoSharp committed:
refactor: SWAMP_DEV.isDev -> SWAMP_DEV.enabled

swampservers commented on lua/autorun/client/cl_devtools.lua:
```diff
@@ -1,144 +1,165 @@
         end
-    end
+        
+        debug.setfenv(loadFunc,environment)
```
Aren't you just setting _G? I don't see why this is necessary

NoSharp replied:
No, it's not just setting _G, as we're loading the environment of the file, the idea is that in the future this could be used to sandbox the code being reloaded to an extent.

swampservers replied:
How is that potentially useful? The point of this code is the ability to test clientside changes to contrib on the live server. The committed code doesn't run with any special environment, so, why would you want to do that?

NoSharp replied:
It was for a potential feature. I can remove it, if you'd like that's fine.

swampservers replied:
Yeah i would say remove it, I am curious what the usefulness of that is though

NoSharp replied:
The intended feature of that is so that you could add a certain layer of sandbox-ing around reloaded scripts, so if you didn't want certain reloaded scripts having access to certain global functions that's the place to do it. This was prior to my understanding of you wanting a hard coded list of allowed users.

swampservers commented on lua/autorun/client/cl_devtools.lua:
```diff
@@ -1,144 +1,165 @@
+        environmentVars = environmentVars or {}
+        loadFunc = CompileString(code, "SwapDevTools/" .. filePath)
+        -- setup our environment
+        local environment = _G
```
Why not just set members of _G?

NoSharp replied:
Addressed in the above comment

swampservers commented on lua/autorun/client/cl_devtools.lua:
```diff
@@ -1,144 +1,165 @@
+    function loadFile(filePath, environmentVars)
+        local code = file.Read(filePath, "MOD")
+        if not code then error("No file exists with the file path: " .. filePath) end
+        environmentVars = environmentVars or {}
```
delete this line and do `for k,v in pairs(environmentVars or {})`

NoSharp replied:
as it's *currently* being iterated over twice, it'd be more optimal to the comparison once than twice in each loop.

swampservers replied:
ah I see. The minute performance difference is totally irrelevant, this isn't a critical section of code and it won't even be running on most clients anyway. But putting that check logic in twice would be redundant.

NoSharp committed:
refactor: swap null check in struct environments

swampservers commented:
> This is a near complete refactor of how the previous iteration of this developer tools utility that was made.
> 
> **Please test this properly before deploying to production. I cannot test this on my single player game.**

You should be able to test it; start up a singleplayer game without contrib installed, and with this script in garrysmod/lua/autorun, then put the contrib folder in place while the game is running?

NoSharp committed:
refactor: swap from GetRank function to allowed list.

NoSharp commented:
That was more in regards to the GetRank check, which is now obsolete.

NoSharp committed:
refactor: use runstring and remove useless sandboxing feature.

NoSharp commented:
That should bring it up to what you'd like, however you'll probably need to fill in steamid's.

NoSharp commented:
bump: anymore on this?

swampservers commented:
Oh sorry I totally missed the notification that this was updated!

swampservers commented on lua/autorun/client/cl_devtools.lua:
```diff
@@ -1,144 +1,171 @@
 
-SWAMP_DEV.structTypes = {
+SWAMP_DEV.structEnvironments = {
     ["weapons"] = function(curClass)
```
I just realized this is missing gamemode (gamemode files use GM, not the same as GAMEMODE)

swampservers commented on lua/autorun/client/cl_devtools.lua:
```diff
@@ -1,144 +1,171 @@
+
+]]--
+
+-- This file is subject to copyright - contact swampservers@gmail.com for more information.
```
this should be first line

swampservers commented on lua/autorun/client/cl_devtools.lua:
```diff
@@ -1,144 +1,171 @@
-SWAMP_DEV.isDev = SWAMP_DEV.isDev
+
+SWAMP_DEV.allowed = {
+    ["SteamID Here."] = true
```
put your own steamid(s) with comment with your nickname

swampservers commented on lua/autorun/client/cl_devtools.lua:
```diff
@@ -1,144 +1,171 @@
-            SWAMP_DEV.structTypes[SWAMP_DEV.itemStruct](string.TrimRight(string.gsub(path, "^.*/", ""), ".lua"))
+local ROOT_DIRECTORY = "addons/contrib/"
+local refreshFolder
+local loadFile
```
what's the point of putting these here and then making a `do ... end`, it's messy

NoSharp replied:
it minimises the scope that gets used and is an optimisation.

swampservers commented on lua/autorun/client/cl_devtools.lua:
```diff
@@ -1,144 +1,171 @@
+        environmentVars = environmentVars or {}
+       
+        -- setup our environment
+        local environment = _G
```
clean this up?

NoSharp replied:
How so?

swampservers replied:
why are you changing the name of this variable instead of just using _G itself? and don't say "for optimization", even if indexing the local variable is 1ns faster that is totally irrelevant in the context of this function which is not critical at all

NoSharp replied:
Its for performance as is everything i do. I focus on optimising everything as much as possible no mattter its calling context.

swampservers replied:
Okay, but, this kind of "optimization" is totally missing the forest because of the trees.

First of all this isn't in a PrePlayerDraw hook. This code only runs _when a developer is using the devtools_ and _they have just edited a file_. That happens probably less than 100 times per day on average, and happens only when the developer is editing something, so we really don't need to give them the extra 0.00001% frames per second this kind of edit would realize (if you generously estimate that this saves 100ns and say they are updating one file every second). Clearly this doesn't make any significant difference, especially compared to something like PrePlayerDraw, which I estimate probably gets run a cumulative 20 billion times per day by our players.

At the same time though this addon _does_ have a significant performance issue in it. The problem is how when checking the file modification times, you check all of the files at once every two seconds rather than amortizing those checks over the two seconds. This is going to spike the frame time every time you check it. I wouldn't be surprised if the spike is quite significant as well, because I assume file.Time requires a synchronous disk read which might take something like 1ms (which adds up over 400 files). I don't notice any issue myself but I have a nvme SSD which is pretty much the best case scenario for that.

I understand the desire to make everything "as optimal as possible", but this is the type of optimization that should probably only be done by a compiler/transpiler if it's wanted. It seems fairly simple to statically analyze a program and localize all repeated global variable accesses, and you might actually get a measurable improvement when doing this kind of optimization across all of the code and in more impactful places. I did this optimization in the gamemode's animations.lua, which is full of hooks that run for every player every tick, and improved the server tickrate by something like 10%. But the context of this function is not impactful at all.

So this change obviously doesn't have any significant upside in performance, but why do I dislike it? It's bloat - it makes the code more confusing. When you see _G getting set, you know you're setting globals, but when you see "environment" getting set, it takes extra analysis to figure out that "environment" is actually just globals. It makes the underlying program/AST more complicated and makes the file longer and take longer to read. Obviously for just this singular change, the negative impact is minor, but what it demonstrates is that you're not trying to make the code as simple as possible. Efficiency doesn't just mean absolute maximum performance when the margin of improvement is so insignificant; optimizing your program to be short and easier to understand is a more important objective.

NoSharp replied:
I understand your response and will remember that for the future.

swampservers commented on lua/autorun/client/cl_devtools.lua:
```diff
@@ -1,144 +1,171 @@
+    
+    local structFound 
+    local ran
+    for fileType in filePath:gmatch("[^/\\]+") do
```
Is there logic here that can be shared with the other refresh function? Maybe just have a flat list of all files, then whenever one is updated, do like RunConsoleCommand("dev_refresh", filename)

NoSharp replied:
Not really as this traverses the path, whereas, in the the automatic refresh it's loading each folder sepeartely.

swampservers commented:
I'll pay you $60 instead since I missed the notification

NoSharp commented:
alright thanks

NoSharp committed:
fix: re-add the GM keyword used by the previous addition

NoSharp committed:
fix: make sure to adjust the environment variable.

swampservers commented:
Thanks, email me how you want payment. I can give you something else to do if you want if you read my comment https://github.com/swampservers/contrib/pull/255#discussion_r658425567

NoSharp commented:
@swampservers I sent you an email with the payment info.
